azimuth_schedule_operator/operator.py

```diff
@@ -46,15 +49,73 @@ async def startup(settings, **kwargs):
 
 
 @kopf.on.cleanup()
-async def cleanup(**kwargs):
+async def cleanup(**_):
     if K8S_CLIENT:
         await K8S_CLIENT.aclose()
     LOG.info("Cleanup complete.")
 
 
-@kopf.on.create(registry.API_GROUP, "schedule")
-@kopf.on.update(registry.API_GROUP, "schedule")
-@kopf.on.resume(registry.API_GROUP, "schedule")
-async def schedule_changed(body, name, namespace, labels, **kwargs):
+async def get_reference(namespace: str, ref: schedule_crd.ScheduleRef):
+    resource = await K8S_CLIENT.api(ref.api_version).resource(ref.kind)
+    object = await resource.fetch(ref.name, namespace=namespace)
+    return object
+
+
+async def delete_reference(namespace: str, ref: schedule_crd.ScheduleRef):
+    resource = await K8S_CLIENT.api(ref.api_version).resource(ref.kind)
+    await resource.delete(ref.name, namespace=namespace)
+
+
+async def check_for_delete(namespace, schedule: schedule_crd.Schedule):
+    # TODO(johngarbutt): add some config in helm to set this?
+    max_delete_duration_int = int(
+        os.environ.get("AZIMUTH_SCHEDULE_MAX_DELETE_DURATION_MINUTES", "15")
+    ) + int(CHECK_INTERVAL_SECONDS / 60)
+    max_delete_duration = datetime.timedelta(minutes=max_delete_duration_int)
+    scheduled_at = schedule.spec.not_after - max_delete_duration
+    now = datetime.datetime.now(datetime.timezone.utc)
+
+    if now >= scheduled_at:
+        LOG.info(f"Attempting delete for {namespace} and {schedule.metadata.name}.")
+        await delete_reference(namespace, schedule.spec.ref)
+        await update_schedule(schedule.metadata.name, namespace, delete_triggered=True)
+    else:
+        LOG.info(f"No delete for {namespace} and {schedule.metadata.name}.")
```

Comment:
I always imagined that this operator would just trigger deletion at the specified time. When we do Blazar, there would be a `Reservation` resource that owns a `Schedule` that would be responsible for ensuring suitable grace.

Reply:
Maybe that should be per platform I guess, the same as the create timeout?

Reply:
OK, yeah, lets delegate this problem to the caller.